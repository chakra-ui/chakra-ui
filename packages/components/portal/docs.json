{
  "PortalManager": {
    "zIndex": {
      "type": "number",
      "required": false,
      "description": "[Z-Index war] If your has multiple elements\nwith z-index clashing, you might need to apply a z-index to the Portal manager"
    }
  },
  "Portal": {
    "appendToParentPortal": {
      "type": "boolean",
      "defaultValue": true,
      "required": false,
      "description": "If `true`, the portal will check if it is within a parent portal\nand append itself to the parent's portal node.\nThis provides nesting for portals.\n\nIf `false`, the portal will always append to `document.body`\nregardless of nesting. It is used to opt out of portal nesting."
    },
    "containerRef": {
      "type": "RefObject<HTMLElement | null>",
      "required": false,
      "description": "The `ref` to the component where the portal will be attached to."
    }
  }
}
