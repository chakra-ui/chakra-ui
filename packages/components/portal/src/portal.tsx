import { useSafeLayoutEffect } from "@chakra-ui/react-use-safe-layout-effect"
import { createContext } from "@chakra-ui/react-context"
import { createPortal } from "react-dom"
import { usePortalManager } from "./portal-manager"
import { useEffect, useMemo, useRef, useState } from "react"

type PortalContext = HTMLDivElement | null

const [PortalContextProvider, usePortalContext] = createContext<PortalContext>({
  strict: false,
  name: "PortalContext",
})

const PORTAL_CLASSNAME = "chakra-portal"
const PORTAL_SELECTOR = `.chakra-portal`

const Container = (props: React.PropsWithChildren<{ zIndex: number }>) => (
  <div
    className="chakra-portal-zIndex"
    style={{
      position: "absolute",
      zIndex: props.zIndex,
      top: 0,
      left: 0,
      right: 0,
      // NB: Don't add `bottom: 0`, it makes the entire app unusable
      // @see https://github.com/chakra-ui/chakra-ui/issues/3201
    }}
  >
    {props.children}
  </div>
)

/**
 * Portal that uses `document.body` as container
 */
const DefaultPortal = (
  props: React.PropsWithChildren<{ appendToParentPortal?: boolean }>,
) => {
  const { appendToParentPortal, children } = props

  const [tempNode, setTempNode] = useState<HTMLElement | null>(null)
  const portal = useRef<HTMLDivElement | null>(null)

  const [, forceUpdate] = useState({})
  useEffect(() => forceUpdate({}), [])

  const parentPortal = usePortalContext()
  const manager = usePortalManager()

  useSafeLayoutEffect(() => {
    if (!tempNode) return

    const doc = tempNode.ownerDocument
    const host = appendToParentPortal ? parentPortal ?? doc.body : doc.body

    if (!host) return

    portal.current = doc.createElement("div")
    portal.current.className = PORTAL_CLASSNAME

    host.appendChild(portal.current)
    forceUpdate({})

    const portalNode = portal.current
    return () => {
      if (host.contains(portalNode)) {
        host.removeChild(portalNode)
      }
    }
  }, [tempNode])

  const _children = manager?.zIndex ? (
    <Container zIndex={manager?.zIndex}>{children}</Container>
  ) : (
    children
  )

  return portal.current ? (
    createPortal(
      <PortalContextProvider value={portal.current}>
        {_children}
      </PortalContextProvider>,
      portal.current,
    )
  ) : (
    <span
      ref={(el) => {
        if (el) setTempNode(el)
      }}
    />
  )
}

interface ContainerPortalProps extends React.PropsWithChildren<{}> {
  containerRef: React.RefObject<HTMLElement | null>
  /**
   * @default false
   */
  appendToParentPortal?: boolean
}

/**
 * Portal that uses a custom container
 */
const ContainerPortal = (props: ContainerPortalProps) => {
  const { children, containerRef, appendToParentPortal } = props
  const containerEl = containerRef.current
  const host =
    containerEl ?? (typeof window !== "undefined" ? document.body : undefined)

  const portal = useMemo(() => {
    const node = containerEl?.ownerDocument.createElement("div")
    if (node) node.className = PORTAL_CLASSNAME
    return node
  }, [containerEl])

  const [, forceUpdate] = useState({})
  useSafeLayoutEffect(() => forceUpdate({}), [])

  useSafeLayoutEffect(() => {
    if (!portal || !host) return
    host.appendChild(portal)
    return () => {
      host.removeChild(portal)
    }
  }, [portal, host])

  if (host && portal) {
    return createPortal(
      <PortalContextProvider value={appendToParentPortal ? portal : null}>
        {children}
      </PortalContextProvider>,
      portal,
    )
  }

  return null
}

export interface PortalProps {
  /**
   * The `ref` to the component where the portal will be attached to.
   */
  containerRef?: React.RefObject<HTMLElement | null>
  /**
   * The content or node you'll like to portal
   */
  children: React.ReactNode
  /**
   * If `true`, the portal will check if it is within a parent portal
   * and append itself to the parent's portal node.
   * This provides nesting for portals.
   *
   * If `false`, the portal will always append to `document.body`
   * regardless of nesting. It is used to opt out of portal nesting.
   *
   * @default true
   */
  appendToParentPortal?: boolean
}

/**
 * Portal
 *
 * Declarative component used to render children into a DOM node
 * that exists outside the DOM hierarchy of the parent component.
 *
 * @see Docs https://chakra-ui.com/portal
 */

export function Portal(props: PortalProps) {
  const portalProps: PortalProps = {
    appendToParentPortal: true,
    ...props,
  }

  const { containerRef, ...rest } = portalProps
  return containerRef ? (
    <ContainerPortal containerRef={containerRef} {...rest} />
  ) : (
    <DefaultPortal {...rest} />
  )
}

Portal.className = PORTAL_CLASSNAME
Portal.selector = PORTAL_SELECTOR

Portal.displayName = "Portal"
