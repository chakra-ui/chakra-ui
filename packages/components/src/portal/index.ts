export { PortalManager, usePortalManager } from "./portal-manager"
export type { PortalManagerProps } from "./portal-manager"
export { Portal } from "./portal"
export type { PortalProps } from "./portal"
