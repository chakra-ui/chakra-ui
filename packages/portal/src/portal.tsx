import { useForceUpdate, useSafeLayoutEffect } from "@chakra-ui/hooks"
import { isBrowser, __DEV__ } from "@chakra-ui/utils"
import { createContext } from "@chakra-ui/react-utils"
import * as React from "react"
import { createPortal } from "react-dom"
import { usePortalManager } from "./portal-manager"

type PortalContext = HTMLDivElement | null

const [PortalContextProvider, usePortalContext] = createContext<PortalContext>({
  strict: false,
  name: "PortalContext",
})

const PORTAL_CLASSNAME = "chakra-portal"
const PORTAL_SELECTOR = `.chakra-portal`

const Container: React.FC<{ zIndex: number }> = (props) => (
  <div
    className="chakra-portal-zIndex"
    style={{
      position: "absolute",
      zIndex: props.zIndex,
      top: 0,
      left: 0,
      right: 0,
      // NB: Don't add `bottom: 0`, it makes the entire app unusable
      // @see https://github.com/chakra-ui/chakra-ui/issues/3201
    }}
  >
    {props.children}
  </div>
)

/**
 * Portal that uses `document.body` as container
 */
const DefaultPortal: React.FC<{ appendToParentPortal?: boolean }> = (props) => {
  const { appendToParentPortal, children } = props

  const tempNode = React.useRef<HTMLDivElement | null>(null)
  const portal = React.useRef<HTMLDivElement | null>(null)

  const forceUpdate = useForceUpdate()

  const parentPortal = usePortalContext()
  const manager = usePortalManager()

  useSafeLayoutEffect(() => {
    if (!tempNode.current) return

    const doc = tempNode.current!.ownerDocument
    const host = appendToParentPortal ? parentPortal ?? doc.body : doc.body

    if (!host) return

    portal.current = doc.createElement("div")
    portal.current.className = PORTAL_CLASSNAME

    host.appendChild(portal.current)
    forceUpdate()

    const portalNode = portal.current
    return () => {
      if (host.contains(portalNode)) {
        host.removeChild(portalNode)
      }
    }
  }, [])

  const _children = manager?.zIndex ? (
    <Container zIndex={manager?.zIndex}>{children}</Container>
  ) : (
    children
  )

  return portal.current ? (
    createPortal(
      <PortalContextProvider value={portal.current}>
        {_children}
      </PortalContextProvider>,
      portal.current,
    )
  ) : (
    <span ref={tempNode} />
  )
}

interface ContainerPortalProps {
  containerRef: React.RefObject<HTMLElement | null>
  appendToParentPortal?: boolean
}

/**
 * Portal that uses a custom container
 */
const ContainerPortal: React.FC<ContainerPortalProps> = (props) => {
  const { children, containerRef, appendToParentPortal } = props
  const containerEl = containerRef.current
  const host = containerEl ?? (isBrowser ? document.body : undefined)

  const portal = React.useMemo(() => {
    const node = containerEl?.ownerDocument.createElement("div")
    if (node) node.className = PORTAL_CLASSNAME
    return node
  }, [containerEl])

  const forceUpdate = useForceUpdate()

  useSafeLayoutEffect(() => {
    forceUpdate()
  }, [])

  useSafeLayoutEffect(() => {
    if (!portal || !host) return
    host.appendChild(portal)
    return () => {
      host.removeChild(portal)
    }
  }, [portal, host])

  if (host && portal) {
    return createPortal(
      <PortalContextProvider value={appendToParentPortal ? portal : null}>
        {children}
      </PortalContextProvider>,
      portal,
    )
  }

  return null
}

export interface PortalProps {
  /**
   * The `ref` to the component where the portal will be attached to.
   */
  containerRef?: React.RefObject<HTMLElement | null>
  /**
   * The content or node you'll like to portal
   */
  children: React.ReactNode
  /**
   * If `true`, the portal will check if it is within a parent portal
   * and append itself to the parent's portal node.
   * This provides nesting for portals.
   *
   * If `false`, the portal will always append to `document.body`
   * regardless of nesting. It is used to opt out of portal nesting.
   */
  appendToParentPortal?: boolean
}

/**
 * Portal
 *
 * Declarative component used to render children into a DOM node
 * that exists outside the DOM hierarchy of the parent component.
 *
 * @see Docs https://chakra-ui.com/portal
 */

export function Portal(props: PortalProps) {
  const { containerRef, ...rest } = props
  return containerRef ? (
    <ContainerPortal containerRef={containerRef} {...rest} />
  ) : (
    <DefaultPortal {...rest} />
  )
}

Portal.defaultProps = {
  appendToParentPortal: true,
}

Portal.className = PORTAL_CLASSNAME
Portal.selector = PORTAL_SELECTOR

if (__DEV__) {
  Portal.displayName = "Portal"
}
