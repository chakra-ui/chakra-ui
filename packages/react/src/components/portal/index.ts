export { Portal } from "./portal"
export type { PortalProps } from "./portal"
