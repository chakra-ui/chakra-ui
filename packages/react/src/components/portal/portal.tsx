import { createPortal } from "react-dom"

export interface PortalProps {
  /**
   * The `ref` to the component where the portal will be attached to.
   */
  containerRef?: React.RefObject<HTMLElement | null>
  /**
   * The content or node you'll like to portal
   */
  children: React.ReactNode
}

/**
 * Portal
 *
 * Declarative component used to render children into a DOM node
 * that exists outside the DOM hierarchy of the parent component.
 *
 * @see Docs https://chakra-ui.com/portal
 */

export function Portal(props: PortalProps) {
  const { containerRef, children } = props
  const mountNode = containerRef?.current ?? globalThis?.document.body
  return mountNode ? createPortal(children, mountNode) : null
}
