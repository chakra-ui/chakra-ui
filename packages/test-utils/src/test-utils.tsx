import { ChakraProvider } from "@chakra-ui/react"
import "@testing-library/jest-dom/extend-expect"
import {
  render as rtlRender,
  RenderOptions,
  fireEvent,
  RenderResult,
} from "@testing-library/react"
import * as React from "react"
import { toHaveNoViolations, axe } from "jest-axe"
import { createSerializer } from "@emotion/jest"
import { RunOptions } from "axe-core"

expect.addSnapshotSerializer(createSerializer())
expect.extend(toHaveNoViolations)

type UI = Parameters<typeof rtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, "wrapper"> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult =>
  rtlRender(
    <ChakraProvider>
      <Wrapper>{ui}</Wrapper>
    </ChakraProvider>,
    options,
  )

export { rtlRender }
export { axe }

export * from "@testing-library/react"

export {
  act as invoke,
  renderHook,
  RenderHookOptions,
  RenderHookResult,
} from "@testing-library/react-hooks"

export { default as userEvent } from "@testing-library/user-event"

export const escape = (ui: HTMLElement) =>
  fireEvent.keyDown(ui, { key: "Escape", keyCode: 27 })

type TestA11YOptions = TestOptions & { axeOptions?: RunOptions }

/**
 * Validates against common a11y mistakes.
 *
 * Wrapper for jest-axe
 *
 * @example
 * ```jsx
 * it('passes a11y test', async () => {
 *  await testA11Y(<MyComponent />, options);
 * });
 *
 * // sometimes we need to perform interactions first to render conditional UI
 * it('passes a11y test when open', async () => {
 *  const { container } = render(<MyComponent />, options);
 *
 *  fireEvent.click(screen.getByRole('button'));
 *
 *  await testA11Y(container, options);
 * });
 * ```
 *
 * @see https://github.com/nickcolley/jest-axe#testing-react-with-react-testing-library
 */
export const testA11y = async (
  ui: UI | Element,
  { axeOptions, ...options }: TestA11YOptions = {},
) => {
  const container = React.isValidElement(ui)
    ? render(ui, options).container
    : ui

  const results = await axe(container, axeOptions)

  expect(results).toHaveNoViolations()
}
