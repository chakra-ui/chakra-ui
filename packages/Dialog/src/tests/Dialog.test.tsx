import * as React from "react"
import { render, userEvent, fireEvent } from "@chakra-ui/test-utils"
import {
  Dialog,
  DialogBody,
  DialogCloseButton,
  DialogContent,
  DialogFooter,
  DialogHeader,
  DialogOverlay,
} from "../Dialog"
import { PortalManager } from "@chakra-ui/portal"

test("Dialog renders correctly", () => {
  const { asFragment } = render(
    <PortalManager>
      <Dialog isOpen onClose={jest.fn()}>
        <DialogOverlay />
        <DialogContent>
          <DialogHeader>Dialog header</DialogHeader>
          <DialogCloseButton />
          <DialogBody>Dialog body</DialogBody>
          <DialogFooter>Dialog footer</DialogFooter>
        </DialogContent>
      </Dialog>
    </PortalManager>,
  )
  expect(asFragment()).toMatchSnapshot()
})

test("has the proper aria attributes", () => {
  const { getByLabelText } = render(
    <PortalManager>
      <Dialog isOpen onClose={jest.fn()}>
        <DialogContent>
          <DialogHeader>Dialog header</DialogHeader>
          <DialogBody>Dialog body</DialogBody>
        </DialogContent>
      </Dialog>
    </PortalManager>,
  )
  // dialog is labelled by the header
  const dialog = getByLabelText("Dialog header")

  expect(dialog).toHaveAttribute("role", "dialog")
  expect(dialog).toHaveAttribute("aria-modal", "true")
  expect(dialog).toHaveAttribute("aria-describedby", "chakra-dialog--body-3")
})

test("clicking the close button calls the onClose callback", () => {
  const onClose = jest.fn()
  const { getByLabelText } = render(
    <PortalManager>
      <Dialog isOpen onClose={onClose}>
        <DialogContent>
          <DialogHeader>Dialog header</DialogHeader>
          <DialogCloseButton />
        </DialogContent>
      </Dialog>
    </PortalManager>,
  )

  // click the close button
  fireEvent.click(getByLabelText("Close"))

  expect(onClose).toHaveBeenCalled()
})

test('clicking overlay or pressing "esc" calls the onClose callback', () => {
  const onClose = jest.fn()
  const { getByTestId } = render(
    <PortalManager>
      <Dialog isOpen onClose={onClose}>
        <DialogOverlay data-testid="overlay" />
        <DialogContent>
          <DialogHeader>Dialog header</DialogHeader>
          <DialogBody>Dialog body</DialogBody>
        </DialogContent>
      </Dialog>
    </PortalManager>,
  )
  const overlay = getByTestId("overlay")

  userEvent.click(overlay)
  fireEvent.keyDown(overlay, { key: "Escape", keyCode: 27 })

  expect(onClose).toHaveBeenCalledTimes(2)
})

test("focuses the initial focus ref when opened", () => {
  const Component = () => {
    const [isOpen, setIsOpen] = React.useState(false)
    const inputRef = React.useRef(null)
    return (
      <PortalManager>
        <button data-testid="button" onClick={() => setIsOpen(true)}>
          Open
        </button>
        <Dialog isOpen={isOpen} initialFocusRef={inputRef} onClose={jest.fn()}>
          <DialogContent>
            <DialogHeader>Dialog header</DialogHeader>
            <DialogBody>
              <input />
              <input />
              <input data-testid="input" ref={inputRef} />
            </DialogBody>
          </DialogContent>
        </Dialog>
      </PortalManager>
    )
  }
  const { getByTestId } = render(<Component />)

  // click button, opening the modal
  fireEvent.click(getByTestId("button"))

  // input is now the active element
  expect(document.activeElement).toEqual(getByTestId("input"))
})

test("returns focus when closed", () => {
  const Component = () => {
    const [isOpen, setIsOpen] = React.useState(false)
    const buttonRef = React.useRef(null)
    return (
      <PortalManager>
        <button
          ref={buttonRef}
          data-testid="button"
          onClick={() => setIsOpen(true)}
        >
          Open
        </button>
        <Dialog
          finalFocusRef={buttonRef}
          isOpen={isOpen}
          onClose={() => setIsOpen(false)}
        >
          <DialogContent>
            <DialogHeader>Dialog header</DialogHeader>
            <DialogCloseButton />
            <DialogBody>Dialog body</DialogBody>
          </DialogContent>
        </Dialog>
      </PortalManager>
    )
  }
  const { getByLabelText, getByTestId } = render(<Component />)
  const button = getByTestId("button")

  // make sure button isn't focused at the start
  expect(document.activeElement).not.toEqual(button)

  // open and close the modal
  fireEvent.click(button)
  fireEvent.click(getByLabelText("Close"))

  expect(document.activeElement).toEqual(button)
})
