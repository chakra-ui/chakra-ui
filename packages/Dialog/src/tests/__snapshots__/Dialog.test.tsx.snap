// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog renders correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-portal-manager"
  >
    <div
      class="chakra-portal"
    >
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="1"
      />
      <div
        data-focus-lock-disabled="false"
      >
        <div
          aria-hidden="true"
          class="css-13vwfqb"
          data-aria-hidden="true"
          data-chakra-dialog-overlay=""
        />
        <section
          aria-describedby="chakra-dialog--body-1"
          aria-labelledby="chakra-dialog--header-1"
          aria-modal="true"
          class="css-1mmk8z2"
          data-chakra-dialog-content=""
          id="chakra-dialog-1"
          role="dialog"
          tabindex="-1"
        >
          <header
            class="css-1kd6u6h"
            data-chakra-dialog-header=""
            id="chakra-dialog--header-1"
          >
            Dialog header
          </header>
          <button
            aria-label="Close"
            class="css-9u5sub"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="css-lhet7y"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                fill="currentColor"
              />
            </svg>
          </button>
          <div
            class="css-5recyo"
            data-chakra-dialog-body=""
            id="chakra-dialog--body-1"
          >
            Dialog body
          </div>
          <footer
            class="css-kz99la"
          >
            Dialog footer
          </footer>
        </section>
      </div>
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
    </div>
  </div>
</DocumentFragment>
`;
